File: daily/Linked List/138-Copy-list-with-random-pointer.go
package daily

type Node struct {
    Val int
    Next *Node
    Random *Node
}

func copyRandomList(head *Node) *Node {
    if head == nil {
        return nil
    }
    
    oldToNew := make(map[*Node]*Node)
    
    // First pass: create copies of all nodes
    curr := head
    for curr != nil {
        oldToNew[curr] = &Node{Val: curr.Val}
        curr = curr.Next
    }
    
    // Second pass: set next and random pointers
    curr = head
    for curr != nil {
        newNode := oldToNew[curr]
        if curr.Next != nil {
            newNode.Next = oldToNew[curr.Next]
        }
        if curr.Random != nil {
            newNode.Random = oldToNew[curr.Random]
        }
        curr = curr.Next
    }
    
    return oldToNew[head]
}

File: daily/Linked List/138-Copy-list-with-random-pointer.py
class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None
        old_node = {}
        curr = head
        while curr:
            old_node[curr] = Node(curr.val)
            curr = curr.next
        
        curr = head
        while curr:
            old_node[curr].next = old_node.get(curr.next)
            old_node[curr].random = old_node.get(curr.random)
            curr = curr.next
        return old_node[head]